### ISSUE TITLE
Generalize NewStreaming[List|Map|Set]->GraphBuilder

### ISSUE THREAD

**Issue by aboodman** [attic-labs/noms#2419]
# Problem

We currently have NewStreamingMap, NewStreamingSet, and NewStreamingList, but we sometimes need to create more complicated structures. For example, nomdex creates `Map<String|Number,Set<T>>`, and we want to modify `csv-import` to be able to create things like `Map<String,Map<String,Map<String,T>>>` (for a 3-column compound key).

My proposal is to generalize `NewStreaming[Map|List|Set]` to support building up complex graphs.

# Proposal

```
type GraphBuilder struct {
  ...
}

// Sets the item identified by |k| in the map at path |p| to |v|.
// |p| can only involve map, set, and list traversal. Structs unsupported.
func (b GraphBuilder) MapSet(p types.Path, k types.Value, v types.Value) GraphBuilder {
...
}

// Appends |v| to the list at path |p|.
// |p| can only involve map, set, and list traversal. Structs unsupported.
func (b GraphBuilder) ListAppend(p types.Path, v types.Value) GraphBuilder {
...
}

// Inserts |v| to the set at path |p|.
// |p| can only involve map, set, and list traversal. Structs unsupported.
func (b GraphBuilder) SetInsert(p types.Path, v types.Value) GraphBuilder {
...
}

// Builds and returns the graph. An error is returned in cases where the graph is malformed, for 
// example, the following returns an error:
//     p := ParsePath("[\"foo\"]")
//     b, err := NewGraphBuilder().MapSet(p, "bar", "baz").ListAppend(p, "hotdog")
// because the implication is that the value at path |p| is both a map and a list.
func (b GraphBuilder) Build() (types.Value, error) {
...
}
```

# Implementation Sketch

The builder would insert into a private leveldb instance (or namespace within a shared instances) key/value pairs, where the key is formed roughly as follows:

- For MapSet() calls, just serialize |p.Append(k)| normally
- For ListAppend() calls, serialize |p| normally, but also add a per-instance unique id or counter
- For SetInsert() calls, serialize |p.Append(v)| normally

There would also have to be a custom comparator so that the keys end up sorted in leveldb in the appropriate way for building maps, sets, and lists quickly. It's possible that a simpler path serialization would be better.

# Future Work

- Extend to allow mutation by adding to the list of supported operations (for example ListSplice(), MapDel()) and passing an existing graph to the constructor.
- Extend to allow building structs by allowing field components in |p|

# Notes

This proposal is slower than just using `NewStreamingList()` in its current form because it writes to leveldb first, then runs through leveldb again to chunk.

In contrast `NewStreamingList` just writes directly to a chunker because in the special case of `NewStreamingList`, the input is sorted by definition.

Unless this can be resolved, we should keep `NewStreamingList` in the codebase to support the case where the input is known to be a sorted list.

**Comment by cmasone-attic** [comment]
I lose you where you say "just serialize `p.Append(k)` normally". `p` is a Path, right? It doesn't have an `Append()` method. Do you mean something like...the levelDB key should be some serialization of the Path, concatenated with some serialization of `k`? And what's the extra unique ID for in the case of lists?

Also, the private leveldb instance that backs the various mutator implementations (the `opCache`) already uses a custom comparator so that inserted values can be enumerated in the best order for making a collection. What's the proposed ordering among collections being build at various paths? Should collections "deeper" in the graph be enumerated first, so that they get built first during Build()?

**Comment by aboodman** [comment]
> I lose you where you say "just serialize p.Append(k) normally". p is a Path, right? It doesn't have an Append() method. Do you mean something like...the levelDB key should be some serialization of the Path, concatenated with some serialization of k?

Sorry, I mean: `append(p, types.IndexPath{k, false}`

> Also, the private leveldb instance that backs the various mutator implementations (the opCache) already uses a custom comparator so that inserted values can be enumerated in the best order for making a collection. What's the proposed ordering among collections being build at various paths? Should collections "deeper" in the graph be enumerated first, so that they get built first during Build()?

I think there has to be at least some custom ordering (or else custom serializing) because you want the path `[52]` to sort before `[421]`.

But I don't think that collections "deeper" need to show up first. I think the only requirement is that things at the same level sort together, and things sort in the order appropriate for efficient building of prolly trees.

**Comment by aboodman** [comment]
> And what's the extra unique ID for in the case of lists?

Just a GUID would be fine. Each call to append is by definition unique. It could also be a global per-process counter.

**Comment by cmasone-attic** [comment]
My problem is that I wasn't sure what it was meant to do :-) I think I see
now: the issue is keeping appends to a List at a given path properly
ordered, right? Because the builder doesn't have "insert at index i", it
just has append. So the required property is that, for a List at a given
path, this ID is ever-increasing, correct?

**Comment by aboodman** [comment]
Oh, I missed "for" in your question.

The GUID is meant to give each call to `ListAppend` a unique key, so that they don't get deduplicated. We can't use the value's ref or anything like that because it's allowed to have duplicate items in a list.

I hadn't thought about the problem of respecting the order `ListAppend()` is called at in at a given level of the graph. So you're right that a global counter would be better.

**Comment by rafael-atticlabs** [comment]
If we're only allowing append for list, it's safe to just use the full path to the list item (e.g. foo[100], foo[101], foo[102]) (i forget the actual path syntax).

**Comment by rafael-atticlabs** [comment]
Also, when we get to mutation of existing graphs, a first step for list would be to limit mutations on a list such that they must be sequential (i.e. splice(1, 0, 2), splice(3, 2, [a, b, c}), etc...). this would make it straight forward to keep mutations sorted.

**Comment by aboodman** [comment]
The reason I didn't use the full path was because I didn't want to have to store the last used index for each list in the graph anywhere (or have to find it via a read). Also, in the future it's more flexible if the keys don't have to be changed to support splices

**Comment by aboodman** [comment]
otoh, if you ever wanted to allow _overwriting_ lists by index, then maybe you'd want to put the index in the key...

**Comment by willhite** [comment]
Now that I've started thinking about this, there are a couple of things I'm not sure of:
1) Using the path object to specify the piece of the graph that is getting built for this api call can be ambiguous. There is no way consistent way to differentiate between a map, set, or list in the path.

2) Does this api require that the entire graph be built, or can parts of the graph be implied just from the path? For example, say I want a nested map of people, where the first map is by State of residence and the inside each of the State maps there is another map keyed by 1st letter of last name. Do each of the State maps need to be explicitly added by a call to MapSet() or is it possible to simple call MapSet(`.states["CA"]`, "W", "John Walters")?

3) It would be nice if any inconsistencies could be caught while the graph was being assembled rather than waiting until the final graph was built.

I believe most of these problems can be solved by passing in a Type to NewGraphBuilder() and using that to do checking and resolve ambiguities but I'm worried that may be too onerous a thing to ask the user to do.

**Comment by willhite** [comment]
New Proposal: GraphBuilder is restricted to building a graph consisting of a hierarchy of Maps with Lists and Sets being possible as leaf nodes in that graph.

The api above changes slightly. Instead of a path object, MapSet(), SetInsert(), and ListAppend() take a []Value that serves as a list of keys for a nested hierarchy of Maps. 

Examples: 
The following pseudo code illustrates how csv-import could build a 3-level map representing a 3-part primary key like so:
```
func readCsv(rows []Row) {
   gb := NewGraphBuilder()
    for i, row in range rows() {
       // getPrimaryKeysFromRow returns []Value with the key of the topmost map being first in slice
        valueSlice := getPrimaryKeysFromRow(row)
        gb.MapSet(pk, row)
    }
    gb.Build()
}
```
Nomdex has a similar structure but is building a graph that consists of a Map<Value, Set<Value>>. It would build a graph using a call to SetInsert():
```
     gb.SetInsert([]Value{idxValue}, row)
```

**Comment by aboodman** [comment]
OK, this looks good.

**Comment by willhite** [comment]
This is done so I'm going to close the issue. One last comment though regarding building a graph that can contain Lists and Sets as nodes in the middle of the graph. I think the best way to do this might be to have some sort of callback from the GraphBuilder.Build() that would take the graphKeys as an argument and get a NomsKind back saying what kind of collection to build.

Fixed by PR #2548
